tpmutil/structures.go

```diff
@@ -14,13 +14,84 @@
 
 package tpmutil
 
-import "io"
+import (
+	"encoding/binary"
+	"io"
+)
 
 // RawBytes is for Pack and RunCommand arguments that are already encoded.
 // Compared to []byte, RawBytes will not be prepended with slice length during
 // encoding.
 type RawBytes []byte
 
+// U16Bytes is a byte slice with a 16-bit header
+type U16Bytes []byte
+
+// TPMMarshal packs U16Bytes
+func (b *U16Bytes) TPMMarshal(out io.Writer) error {
+	size := uint16(len([]byte(*b)))
+	if err := binary.Write(out, binary.BigEndian, size); err != nil {
+		return err
+	}
+	if err := binary.Write(out, binary.BigEndian, []byte(*b)); err != nil {
+		return err
+	}
+	return nil
+}
+
+// TPMUnmarshal unpacks a U16Bytes
+func (b *U16Bytes) TPMUnmarshal(in io.Reader) error {
+	var tmpSize uint16
+	if err := binary.Read(in, binary.BigEndian, &tmpSize); err != nil {
+		return err
+	}
+	size := int(tmpSize)
+	if len([]byte(*b)) >= size {
+		*b = (*b)[:size]
+	} else {
+		*b = append(*b, make([]byte, size-len(*b))...)
+	}
+
+	if err := binary.Read(in, binary.BigEndian, b); err != nil {
```

Comment:
can this be `in.Read([]byte(*b))` ?

Reply:
it could be `if _, err := in.Read([]byte(*b)); err != nil' yeah

Reply:
Right, just so that we don't incur the type switching overhead of binary.Read

Reply:
Could also switch to using out.Write above.